numpy/polynomial/polyutils.py

```diff
@@ -415,51 +416,87 @@ def mapdomain(x, old, new):
 
-def _vander3d(vander_f, x, y, z, deg):
-    """
-    Helper function used to implement the ``<type>vander3d`` functions.
+    .. math::
+        W[i_0, \ldots, i_M, j_0, \ldots, j_N] = \prod_{k=0}^N{V_k(x_k)[i_0, \ldots, i_M, j_k]}
+
+    where
+
+    .. math::
+        N &= \texttt{len(points)} = \texttt{len(degrees)} = \texttt{len(vander\_fs)} \\
+        M &= \texttt{points[k].ndim} \\
+        V_k &= \texttt{vander\_fs[k]} \\
+        x_k &= \texttt{points[k]} \\
+        0 \le j_k &\le \texttt{degrees[k]}
+
+    Expanding the one-dimensional :math:`V_k` functions gives:
+
+    .. math::
+        W[i_0, \ldots, i_M, j_0, \ldots, j_N] = \prod_{k=0}^N{B_{k, j_k}(x_k[i_0, \ldots, i_M])}
+
+    where :math:`B_{k,m}` is the m'th basis of the polynomial construction used along
+    dimension :math:`k`. For a regular polynomial, :math:`B_{k, m}(x) = P_m(x) = x^m`.
 
     Parameters
     ----------
-    vander_f : function(array_like, int) -> ndarray
-        The 1d vander function, such as ``polyvander``
-    x, y, z, deg :
-        See the ``<type>vander3d`` functions for more detail
-    """
-    degx, degy, degz = [
-        _deprecate_as_int(d, "degrees")
-        for d in deg
-    ]
-    x, y, z = np.array((x, y, z), copy=False) + 0.0
-
-    vx = vander_f(x, degx)
-    vy = vander_f(y, degy)
-    vz = vander_f(z, degz)
-    v = vx[..., None, None]*vy[..., None,:, None]*vz[..., None, None,:]
-    return v.reshape(v.shape[:-3] + (-1,))
+    vander_fs : Sequence[function(array_like, int) -> ndarray]
+        The 1d vander function to use for each axis, such as ``polyvander``
+    points : Sequence[array_like]
+        Arrays of point coordinates, all of the same shape. The dtypes
+        will be converted to either float64 or complex128 depending on
+        whether any of the elements are complex. Scalars are converted to
+        1-D arrays.
+        This must be the same length as `vander_fs`.
+    degrees : Sequence[int]
+        The maximum degree (inclusive) to use for each axis.
+        This must be the same length as `vander_fs`.
+
+    Returns
+    -------
+    vander_nd : ndarray
+        An array of shape ``points[0].shape + tuple(d + 1 for d in degrees)``.
+    """
+    n_dims = len(vander_fs)
+    if n_dims != len(points):
+        raise ValueError(
+            "Expected {} dimensions of sample points, got {}".format(n_dims, len(points)))
+    if n_dims != len(degrees):
+        raise ValueError(
+            "Expected {} dimensions of degrees, got {}".format(n_dims, len(degrees)))
+    if n_dims == 0:
+        raise ValueError("Unable to guess a dtype or shape when no points are given")
+
+    # convert to the same shape and type
+    points = tuple(
+        np.array(tuple(points), copy=False) + 0.0
+    )
+    # we checked this wasn't empty already, so no `initial` needed
+    return functools.reduce(operator.mul, (
+        # slicing inserts a dimension per degree
+        vander_f(p, deg)[(...,) + _nth_slice(i, n_dims)]
```

Comment:
It might be clearer to just put this in a `for` loop, I doubt there would be much difference in speed.

Reply:
Putting this in a for loop  ends up as something messy like:
```python
result = None
for ... in zip(...):
    part = vander_f(p, deg)[(...,) + _nth_slice(i, n_dims)
    if result is not None:
        result += part
    else:
        result = part
assert result is not None
```

Reply:
That doesn't look right. Something like
```
aslice = (slice(None),)
result = vander_fs[0](points[0], degrees[0])
for i in range(1, n_dims):
    aslice = (None,) + aslice
    result = result[..., None] * vander_fs[i](points[i], degs[i])[(...,) + aslice]
return result
```
Returning the result of a function call with a complicated set of arguments just does not read well and I don't think worth the use of reduce. There are lots of ways to break up the steps.

Reply:
IMO incrementally slicing `result` is more confusing - your loop replaces a simple reduce operation  (multiply) with a complex one (slice and multiply)

How about I split my version over multiple lines:
```python
# produce the vander-monde matrix for each dimension, placing each of the trailing axes in a unique location
vander_arrays = (
    vander_f(p, deg)[(...,) + _nth_slice(i, n_dims)]
    for i, (vander_f, p, deg) in enumerate(zip(vander_fs, points, degrees))
)

return functools.reduce(operator.mul, vander_arrays)
```

Reply:
I like the split versions better, with a slight (0+) preference for the second. Maybe indent the closing parenthesis.

Note that that incremental slicing has fewer operations and one less function to track down, but don't care much one way or the other. It is like an inductive proof :)

Reply:
I'm not a huge fan of induction unless its necessary.

Either way - mind reviewing gh-14352 before I update this one?

Reply:
> Note that that incremental slicing has fewer operations

You have two slices per access, whereas I only have one. Admittedly your slices are cheaper to construct though. I doubt it matters, and if it did `_nth_slice` could become a lookup table.